libs/model/src/contest/GetAllContests.query.ts

```diff
@@ -1,15 +1,75 @@
-import { Query } from '@hicommonwealth/core';
+import { Query, logger } from '@hicommonwealth/core';
 import * as schemas from '@hicommonwealth/schemas';
 import { QueryTypes } from 'sequelize';
+import { fileURLToPath } from 'url';
 import { z } from 'zod';
 import { models } from '../database';
+import { rollOverContest } from '../services/commonProtocol/contestHelper';
+
+const __filename = fileURLToPath(import.meta.url);
+const log = logger(__filename);
+
+// find community contests that are ended and trigger rollover
+async function performContestRollovers(communityId: string) {
+  const contestManagersWithEndedContest = await models.sequelize.query<{
+    contest_address: string;
+    interval: number;
+    url: string;
+  }>(
+    `
+    SELECT cm.contest_address, cm.interval, co.contest_id, co.end_time, COALESCE(cn.private_url, cn.url) as url
+    FROM "ContestManagers" cm
+    JOIN (
+        SELECT *
+        FROM "Contests"
+        WHERE (contest_address, contest_id) IN (
+            SELECT contest_address, MAX(contest_id) AS contest_id
+            FROM "Contests"
+            GROUP BY contest_address
+        )
+    ) co ON co.contest_address = cm.contest_address AND NOW() > co.end_time
+    JOIN "Communities" cu ON cm.community_id = cu.id AND cu.id = :community_id
+    JOIN "ChainNodes" cn ON cu.chain_node_id = cn.id;
+  `,
+    {
+      type: QueryTypes.SELECT,
+      raw: true,
+      replacements: { community_id: communityId },
+    },
+  );
+
+  const contestRolloverPromises = contestManagersWithEndedContest.map(
+    async ({ url, contest_address, interval }) => {
+      log.debug(`ROLLOVER: ${contest_address}`);
+      return rollOverContest(url, contest_address, interval === 0);
+    },
+  );
+
+  const promiseResults = await Promise.allSettled(contestRolloverPromises);
+
+  const errors = promiseResults
+    .filter(({ status }) => status === 'rejected')
+    .map(
+      (result) =>
+        (result as PromiseRejectedResult).reason || '<unknown reason>',
+    );
+
+  if (errors.length > 0) {
+    log.warn(
+      `GetAllContests updateScore: failed with errors: ${errors.join(', ')}"`,
+    );
+  }
+}
 
 export function GetAllContests(): Query<typeof schemas.GetAllContests> {
   return {
     ...schemas.GetAllContests,
     auth: [],
     secure: false,
     body: async ({ payload }) => {
+      // do in background
+      performContestRollovers(payload.community_id);
```

Comment:
Since it's an onchain transaction in which the system needs to wait for the event to be processed, the rollover won't be "complete" until a few seconds later. So there's no use in waiting for it here. But this also means that the user may need to refresh again to see the result of the rollover. Definitely a temporary hack solution.

Reply:
it takes around 1min 15sec to restart the contest

https://github.com/hicommonwealth/commonwealth/assets/14819225/803aa1c6-9f33-4c34-884f-b0dd22ada8ea